nixos/modules/services/matrix/synapse.nix

```diff
@@ -784,64 +988,123 @@ in {
-        RestrictNamespaces = true;
-        RestrictRealtime = true;
-        RestrictSUIDSGID = true;
-        SystemCallArchitectures = "native";
-        SystemCallFilter = [ "@system-service" "~@resources" "~@privileged" ];
-      };
+    };
+
+    systemd.services =
+      let
+        targetConfig =
+          if hasWorkers
+          then {
+            partOf = [ "matrix-synapse.target" ];
+            wantedBy = [ "matrix-synapse.target" ];
+            unitConfig.ReloadPropagatedFrom = "matrix-synapse.target";
+          }
+          else {
+            after = [ "network.target" ] ++ optional hasLocalPostgresDB "postgresql.service";
+            wantedBy = [ "multi-user.target" ];
+          };
+        baseServiceConfig = {
+          environment = optionalAttrs (cfg.withJemalloc) {
+            LD_PRELOAD = "${pkgs.jemalloc}/lib/libjemalloc.so";
+          };
+          serviceConfig = {
+            Type = "notify";
+            User = "matrix-synapse";
+            Group = "matrix-synapse";
+            WorkingDirectory = cfg.dataDir;
+            ExecReload = "${pkgs.util-linux}/bin/kill -HUP $MAINPID";
+            Restart = "on-failure";
+            UMask = "0077";
+
+            # Security Hardening
+            # Refer to systemd.exec(5) for option descriptions.
+            CapabilityBoundingSet = [ "" ];
+            LockPersonality = true;
+            NoNewPrivileges = true;
+            PrivateDevices = true;
+            PrivateTmp = true;
+            PrivateUsers = true;
+            ProcSubset = "pid";
+            ProtectClock = true;
+            ProtectControlGroups = true;
+            ProtectHome = true;
+            ProtectHostname = true;
+            ProtectKernelLogs = true;
+            ProtectKernelModules = true;
+            ProtectKernelTunables = true;
+            ProtectProc = "invisible";
+            ProtectSystem = "strict";
+            ReadWritePaths = [ cfg.dataDir ];
+            RemoveIPC = true;
+            RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_UNIX" ];
+            RestrictNamespaces = true;
+            RestrictRealtime = true;
+            RestrictSUIDSGID = true;
+            SystemCallArchitectures = "native";
+            SystemCallFilter = [ "@system-service" "~@resources" "~@privileged" ];
+          };
+        }
+        // targetConfig;
+        genWorkerService = name: workerCfg:
+          let
+            finalWorkerCfg = workerCfg // { worker_name = name; };
+            workerConfigFile = format.generate "worker-${name}.yaml" finalWorkerCfg;
+          in
+          {
+            name = "matrix-synapse-worker-${name}";
+            value = lib.mkMerge [
+              baseServiceConfig
+              {
+                description = "Synapse Matrix worker ${name}";
+                # make sure the main process starts first for potential database migrations
+                after = [ "matrix-synapse.service" ];
+                serviceConfig = {
```

Comment:
https://github.com/NixOS/nixpkgs/pull/242026/files#diff-36428133f6120f9176e7210bcb4f75292a0c9cab54235840b13ba961b453d6d4R167-R171

This was never resolved and is relevant for this implementation as well

Reply:
Can you give me some instructions to reproduce the underlying issue? Because I cannot reproduce this at all.

Example 1 (`systemctl restart matrix-synapse.target`):

```
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix worker worker0...
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix worker worker1...
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix worker worker2...
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker1.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix worker worker1.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker1.service: Consumed 14min 23.993s CPU time, read 281.0K from disk, written 0B to disk, received 47.5M IP traffic, sent 101.7M IP traffic.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker2.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix worker worker2.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker2.service: Consumed 13min 42.896s CPU time, read 785.5K from disk, written 0B to disk, received 46.9M IP traffic, sent 94.6M IP traffic.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker0.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix worker worker0.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker0.service: Consumed 26min 15.411s CPU time, read 3.3M from disk, written 0B to disk, received 354.3M IP traffic, sent 165.7M IP traffic.
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix homeserver...
Sep 11 19:39:33 soost systemd[1]: matrix-synapse.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix homeserver.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse.service: Consumed 45min 50.242s CPU time, read 78.9M from disk, written 0B to disk, received 153.6M IP traffic, sent 343.5M IP traffic.
Sep 11 19:39:33 soost systemd[1]: Starting Synapse Matrix homeserver...
Sep 11 19:39:44 soost systemd[1]: Started Synapse Matrix homeserver.
Sep 11 19:39:44 soost systemd[1]: Starting Synapse Matrix worker worker0...
Sep 11 19:39:44 soost systemd[1]: Starting Synapse Matrix worker worker1...
Sep 11 19:39:44 soost systemd[1]: Starting Synapse Matrix worker worker2...
Sep 11 19:39:49 soost systemd[1]: Started Synapse Matrix worker worker1.
Sep 11 19:39:50 soost systemd[1]: Started Synapse Matrix worker worker0.
Sep 11 19:39:50 soost systemd[1]: Started Synapse Matrix worker worker2.
```

Everything starts up properly most notably the workers start up in parallel (notice `Starting` vs `Started` _after_ synapse itself).

Same for the case where I restart all services individually:

```
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix homeserver...
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix worker worker0...
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix worker worker1...
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix worker worker2...
Sep 11 19:42:10 soost systemd[1]: matrix-synapse-worker-worker0.service: Deactivated successfully.
Sep 11 19:42:10 soost systemd[1]: Stopped Synapse Matrix worker worker0.
Sep 11 19:42:10 soost systemd[1]: matrix-synapse-worker-worker0.service: Consumed 8.753s CPU time, received 573.3K IP traffic, sent 353.0K IP traffic.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker1.service: Deactivated successfully.
Sep 11 19:42:11 soost systemd[1]: Stopped Synapse Matrix worker worker1.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker1.service: Consumed 9.644s CPU time, received 328.4K IP traffic, sent 400.9K IP traffic.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker2.service: Deactivated successfully.
Sep 11 19:42:11 soost systemd[1]: Stopped Synapse Matrix worker worker2.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker2.service: Consumed 9.393s CPU time, received 308.8K IP traffic, sent 361.6K IP traffic.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse.service: Deactivated successfully.
Sep 11 19:42:11 soost systemd[1]: Stopped Synapse Matrix homeserver.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse.service: Consumed 30.249s CPU time, received 985.8K IP traffic, sent 1.1M IP traffic.
Sep 11 19:42:11 soost systemd[1]: Starting Synapse Matrix homeserver...
Sep 11 19:42:19 soost systemd[1]: Started Synapse Matrix homeserver.
Sep 11 19:42:19 soost systemd[1]: Starting Synapse Matrix worker worker0...
Sep 11 19:42:19 soost systemd[1]: Starting Synapse Matrix worker worker1...
Sep 11 19:42:19 soost systemd[1]: Starting Synapse Matrix worker worker2...
Sep 11 19:42:25 soost systemd[1]: Started Synapse Matrix worker worker0.
Sep 11 19:42:26 soost systemd[1]: Started Synapse Matrix worker worker2.
Sep 11 19:42:26 soost systemd[1]: Started Synapse Matrix worker worker1.
```

Reply:
I have conflicting information about this: On one hand this _should_ not be necessary, but in real deployments we have noticed that without it synapse workers will fail whenever there's a schema upgrade and you run `nixos-rebuild switch`

Solutions as I see it is either that script, or we could change some of the systemd startLimit stuff to either make it wait longer between restarts (so it doesn't reach it's burstLimit) or adjust the burst limits so that it just keeps trying.

I prefer the wait for script

Alternatively we could assume we're wrong in our observations and that this is a non-issue, but I suspect we're going to end up with adding something similar here in a couple of months then anyways.

Reply:
Do you still remember when was the last time you've observed it? Perhaps it got fixed in systemd since then (would check the NEWS from it, but I'd like to know where to start approximately).

Reply:
The commit where it was added to the module this was ported from was Oct 10, 2022 so probably then at least.

It could also be funky stuff with nixos-rebuild generating new service files with a pkg-upgrade or something. I recall it being hard to reproduce

Reply:
> It could also be funky stuff with nixos-rebuild generating new service files with a pkg-upgrade or something

@dali99 If `switch-to-configuration.pl` replaces units (and they have differences) a restart (in some cases a reload, but that's not relevant here) will be done. I'm very sure that this happens within the same transaction. The only exception would be if you have things that get stopped and started later on whereas others get restarted, that happens in multiple transactions IIRC, but this doesn't seem to be the case here.

@999eagle you're running this module for the longest time I guess. Just wondering do you remember any issues with synapse (with workers) getting up again after a deployment?

----

My problem is that right now everything seems to indicate that this isn't a problem anymore and systemd with the config we have in this PR works fine. I guess I'll ask in the nixos systemd channel if someone who probably has more inside knowledge about systemd can take a look and share their opinion, then we'll see how to proceed.

Reply:
So, while I'm rather certain that the ordering stuff is fine here(Type=notify, `sd_notify(READY=1)` in synapse main happens _after_ successful schema migrations, restart of workers+synapse happens in a single transaction in switch-to-configuration.pl, some of the moving parts here might've changed over time), I don't want to discard the reports entirely.

First of all I've requested in the nixos/systemd channel on Matrix to take a look here, perhaps I've missed something.

After chatting with @mweinelt a bit, we agreed on the following approach:

* We'll observe this for a bit (i.e. await the next synapse bump with schema changes). We have at least two testers (@999eagle and me), though I'm happy for more folks to try this out!
* If we don't encounter any issues with workers failing to start up, we'll merge this into unstable.
* If there's still a case left that'll fail, there's a good chance that we'll catch this before the next branch-off.

My problem is that if we add this now, we'll have a bunch of code which is most likely a no-op (but as I said I may be wrong) that is hard to remove because we _think_ that it might cause issues that are hard to reproduce. That's why I'd like to have this cleared up now.

Reply:
> do you remember any issues with synapse (with workers) getting up again after a deployment?
> under which circumstances did the workers actually crash?

I think I had one crash like this a few months back when a new Synapse version required a database schema upgrade. Workers actually failed to start because the main process hadn't upgraded the database yet.

I haven't observed that behavior in quite a while though (despite multiple database schema upgrades happening), so it's either been fixed in some way by Synapse (probably by moving the `sd_notify` to happen after the schema upgrade is done) or it's a really weird edge case that's very hard to reproduce. I'm hoping it's the former, making a workaround for this no longer required.

Reply:
I'm also on board with testing, since I've just migrated to this PR.

Migrating from 1.89 to 1.91.2 forced a schema migration on the main process and during the switch-to-configuration all synapse units were stopped and started again. The order of units is not even honored inside a single transaction from what @Ma27 and me could see. While colmena told me the switch had failed, the units however recovered quickly and everything came up as needed.

Reply:
>While colmena told me the switch had failed, the units however recovered quickly

This is indicative of it just restarting the units after failure instead of waiting.

There might be timing related issues where depending on the size of the database, or just the size of the migration The systemd timeout will get reached. And it will stop trying to restart the worker

Reply:
I will turn off my wait-for-synapse script as well to help with the datagathering

Reply:
> The order of units is not even honored inside a single transaction from what @Ma27 and me could see

It's slightly weird that this didn't happen on my experiments.

But anyways, for reasons I have enough of synapse for today. @mweinelt has a setup where the issue can be observed and there's a new release out with a schema change, so there are plenty of data points now for me to investigate. Will try to do that tomorrow, then we'll see how to proceed.

Reply:
I've updated my Synapse instance to 1.91 and actually ran into this issue again. Workers were started before database migration was done and immediately exited with a failure. Simply starting them again fixed the issue though. The underlying issue with unit dependencies not being handled correctly is most likely an issue with either `systemd` itself or `switch-to-configuration.pl`.
To fix this for `matrix-synapse` I think the cleaner solution would be to add `RestartSec=1s` to the services and maybe set `Restart=always` instead of `Restart=on-failure`. This would allow for easier overriding for users of this module and would cause systemd to handle the wait before automatically restarting the worker services instead of having to handle this manually in a script.

Reply:
The issue is that, as far as I remember it, systemd only guarantees dependency resolution and ordering during system bring-up, not when you call into `systemctl restart matrix-synapse.service matrix-synapse-worker-foo.service matrix-synapse-worker-bar.service`. They would all be started at the same time. And that is exactly where `switch-to-configuration.pl` comes up short.

Reply:
As far as I can tell, services are actually stopped and started instead of restarted during a switch. So `systemd` is apparently ignoring unit dependencies even during `systemctl start`. Another option to work around this might be to start `matrix-synapse.target` instead of all services individually as `systemd` does respect unit dependencies then.
Can we make `switch-to-configuration.pl` stop all services and the target and then start only the target?

Reply:
In the same [commit](https://github.com/dali99/nixos-matrix-modules/commit/ccbd09267948d5f51ed0fbb1d799a1aa63c4bdbf) as I added the wait-for-script I also added a requires on matrix-synapse to the workers like you did in 24f6a70 so maybe this is indeed enough.

I've been running my module with only the `requires` since last week and it didn't fail with upgrades during that time, for some anecdotal evidence

Reply:
Just deployed 1.92.1 with @999eagle's latest commit applied and it looks good to me:

```
Sep 18 15:48:39 soost systemd[1]: Starting Synapse Matrix homeserver...
Sep 18 15:48:49 soost systemd[1]: Started Synapse Matrix homeserver.
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker client-0...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker client-1...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-receiver-0...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-receiver-1...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-sender-0...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-sender-1...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker misc...
Sep 18 15:48:55 soost systemd[1]: Started Synapse Matrix worker federation-sender-1.
Sep 18 15:48:55 soost systemd[1]: Started Synapse Matrix worker client-0.
Sep 18 15:48:55 soost systemd[1]: Started Synapse Matrix worker federation-receiver-0.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker federation-sender-0.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker misc.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker federation-receiver-1.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker client-1.
```

This time not only via `systemctl restart`, but with `switch-to-configuration.pl` (i..e an actual deploy).

> I've been running my module with only the requires since last week and it didn't fail with upgrades during that time, for some anecdotal evidence

While I'm not 100% convinced by the explanation about how systemd falls short here, I don't think that matters anymore. We have three confirmations that the current behavior is fine now (and what I'd consider desirable).

Even if this falls short, we have a restart behavior in place that will catch that. The only case would be a database that's so slow to migrate that the start timeout (or in now unlikely cases the restart threshold if the dependency handling actually breaks) gets exceeded. However I think that in such a case it's the job of an administrator to increase the timeouts accordingly. Also, I think that restart=always with a configuration that will never get into the StartLimit Burst (i.e. that systemd gives up on start/restart) is not a good idea since this will break monitoring that relies on the failed state of a service (e.g. prometheus-node-exporter alerts) because the failed state will only be reached for a very short period of time.

All in all, I think we can resolve this now.